include/boost/math/statistics/detail/single_pass.hpp

```diff
@@ -0,0 +1,384 @@
+    static unsigned num_threads {std::thread::hardware_concurrency()};
+
+    const auto elements {std::distance(first, last)};
+    const auto range_a {std::floor(elements / 2)};
+    const auto range_b {elements - range_a};
+
+    thread_counter.fetch_add(2);
+    auto future_a {std::async(std::launch::async, [first, range_a]() -> ReturnType
+    {
+        if(thread_counter + 2 <= num_threads && range_a > 10)
+        {
+            return parallel_variance_impl<ReturnType>(first, std::next(first, range_a));
+        }
+        else
+        {
+            return variance_sequential_impl<ReturnType>(first, std::next(first, range_a));
+        }
+    })};
+    auto future_b {std::async(std::launch::async, [first, last, range_a]() -> ReturnType
+    {
+        if(thread_counter + 2 <= num_threads && range_a > 10)
+        {
+            return parallel_variance_impl<ReturnType>(std::next(first, range_a), last);
+        }
+        else
+        {
+            return variance_sequential_impl<ReturnType>(std::next(first, range_a), last);
+        }
+    })};
+
+    const auto results_a {future_a.get()};
+    const auto results_b {future_b.get()};
+
+    const auto mean_a = std::get<0>(results_a);
+    const auto M2_a = std::get<1>(results_a);
+    const auto mean_b = std::get<0>(results_b);
+    const auto M2_b = std::get<1>(results_b);
+
+    const auto n_ab = elements;
+    const auto delta = mean_b - mean_a;
+    const auto mean_ab = (range_a * mean_a + range_b * mean_b) / n_ab;
+    const auto M2_ab = M2_a + M2_b + delta * delta * (range_a * range_b / n_ab);
+
+    return std::make_tuple(mean_ab, M2_ab, n_ab);
+}
+
+// https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Higher-order_statistics
+template<typename ReturnType, typename ForwardIterator>
+ReturnType first_four_moments_sequential_impl(ForwardIterator first, ForwardIterator last)
+{
+    using Real = std::tuple_element_t<0, ReturnType>;
+    using Size = std::tuple_element_t<4, ReturnType>;
+
+    Real M1 = *first;
+    Real M2 = 0;
+    Real M3 = 0;
+    Real M4 = 0;
+    Size n = 2;
+    for (auto it = std::next(first); it != last; ++it)
+    {
+        Real delta21 = *it - M1;
+        Real tmp = delta21/n;
+        M4 = M4 + tmp*(tmp*tmp*delta21*((n-1)*(n*n-3*n+3)) + 6*tmp*M2 - 4*M3);
+        M3 = M3 + tmp*((n-1)*(n-2)*delta21*tmp - 3*M2);
+        M2 = M2 + tmp*(n-1)*delta21;
+        M1 = M1 + tmp;
+        n += 1;
+    }
+
+    return std::make_tuple(M1, M2, M3, M4, n-1);
+}
+
+template<typename ReturnType, typename ForwardIterator>
+ReturnType parallel_first_four_moments_impl(ForwardIterator first, ForwardIterator last)
+{
+    using Real = typename std::iterator_traits<ForwardIterator>::value_type;
+    static unsigned num_threads {std::thread::hardware_concurrency()};
+
+    const auto elements {std::distance(first, last)};
+    const auto range_a {std::floor(elements / 2)};
+    const auto range_b {elements - range_a};
+
+    thread_counter.fetch_add(2);
```

Comment:
Looks like you're using the same global static thread counter for this as you do for the variance, which could lead to race conditions.

Reply:
The thread counter and `fetch_add` are both atomic.

Reply:
Just because the counter is atomic doesn't mean race conditions are impossible. The counter can become incorrect. For example, try launching two threads running this algorithm and the variance at the same time. Each will increment the atomic counter, but then read a value that is incorrect.

Reply:
In the event that you run this and variance simultaneously the number of threads will be synced globally to avoid over-subscription. The latest commit offers better synchronization.